test/custom_task_test.go

```diff
@@ -401,3 +403,130 @@ spec:
+	knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+	defer tearDown(ctx, t, c, namespace)
+
+	// Create a custom task controller
+	applyController(t)
+	// Cleanup the controller after finishing the test
+	defer cleanUpController(t)
+
+	for _, tc := range []struct {
+		name                string
+		duration            string
+		conditionAccessorFn func(string) ConditionAccessorFn
+		wantConditionType   apis.ConditionType
+		wantConditionStatus corev1.ConditionStatus
+		wantConditionReason string
+	}{{
+		name:                "Wait Task Has Passed",
+		duration:            "1s",
+		conditionAccessorFn: Succeed,
+		wantConditionType:   apis.ConditionSucceeded,
+		wantConditionStatus: corev1.ConditionTrue,
+		wantConditionReason: "DurationElapsed",
+	}, {
+		name:                "Wait Task Is Waiting",
+		duration:            "2s",
+		conditionAccessorFn: Running,
+		wantConditionType:   apis.ConditionSucceeded,
+		wantConditionStatus: corev1.ConditionUnknown,
+		wantConditionReason: "Waiting",
+	}} {
+		t.Run(tc.name, func(t *testing.T) {
+			runName := helpers.ObjectNameForTest(t)
+			run := parse.MustParseRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+`, runName, apiVersion, kind, tc.duration))
+			if _, err := c.RunClient.Create(ctx, run, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Failed to create TaskRun %q: %v", runName, err)
+			}
+
+			// Wait for the Run.
+			if err := WaitForRunState(ctx, c, runName, time.Minute, tc.conditionAccessorFn(runName), tc.wantConditionReason); err != nil {
+				t.Fatalf("Waiting for Run to finish running: %v", err)
+			}
+
+			// Get the actual Run
+			gotRun, err := c.RunClient.Get(ctx, runName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("%v", err)
+			}
+
+			gotCondition := gotRun.GetStatus().GetCondition(apis.ConditionSucceeded)
+			if gotCondition == nil {
+				t.Fatal("The Run failed to succeed")
+			}
+
+			// Compose the expected Run
+			runYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  ref:
+    apiVersion: %s
+    kind: %s
+  params:
+  - name: duration
+    value: %s
+  serviceAccountName: default
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 1
+  `, runName, namespace, apiVersion, kind, tc.duration, tc.wantConditionReason, tc.wantConditionStatus, tc.wantConditionType)
+			wantRun := parse.MustParseRun(t, runYAML)
+			var (
+				filterTypeMeta   = cmpopts.IgnoreFields(v1alpha1.Run{}, "TypeMeta")
+				filterObjectMeta = cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ResourceVersion", "UID", "CreationTimestamp", "Generation", "ManagedFields")
+				filterCondition  = cmpopts.IgnoreFields(apis.Condition{}, "LastTransitionTime.Inner.Time", "Message")
+				filterRunStatus  = cmpopts.IgnoreFields(v1alpha1.RunStatusFields{}, "StartTime", "CompletionTime")
```

Comment:
could this be moved to the top level so that they can be reused in other end to end tests?

Reply:
Totally agree that this is a great idea after we add more test cases using those vars in other funcs. For now I'm trying to follow https://wiki.c2.com/?DeclareVariablesAtFirstUse to declare it where it is first used. wdyt?

Reply:
sounds good 👍